Social Icons Widget: Add more common feed url components to supported icons

rickcurran opened the pull request:
As an improvement based on the issue #10662 "Social Icons Widget: allow for alternate feed URLs" this adds some more common feed url components to the get_supported_icons() function. The existing version only matches the existence of "/feed/" in the url being added, as such many other feed URL formats would end up showing with the default link icon instead. Whilst it may still not catch all possible feed url formats it will catch many commonly used formats as well as non-permalink versions of WordPress feeds.

This feature could be tested by beta testers to see if more feed urls are correctly recognised and shown with the RSS feed icon instead of the generic link icon.

Fixes #10662

#### Changes proposed in this Pull Request:
* This PR adds in approximately 20 additional items to an array `$social_links_icons` which is used to match components of any url added to the Social Icons widget, the additions expand upon the existing `/feed/` array item which is intended to show an RSS feed icon if matched. Whilst this works for the standard WordPress permalink feed format it doesn't work for non-permalink WordPress feed urls and also many other common RSS feed url formats.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds features to the `get_supported_icons()` function which matches urls added to the Social Icons widget to specific icons for the URL. This addition provides more items to the `$social_links_icons` array which match other common RSS feed url structures.

#### Testing instructions:
To test this please try adding additional RSS feed urls to the Social Icons widget, the standard WordPress url for `somedomain.com/feed/` is already supported so try some more unusual formats such as those generated by other CMS or blogging platforms. Here are some suggested urls to try:
* http://blogs.adobe.com/jkost/atom.xml
* http://feeds.feedburner.com/clickontyler
* http://daringfireball.net/index.xml
* http://ranchero.com/xml/rss.xml
* http://usefulmac.com/rss
* http://rss.macworld.com/macworld/feeds/main

If these are correctly parsed then you should see an RSS feed icon instead of the default 'chain link' icon.

#### Proposed changelog entry for your changes:
* This improves the likelihood that RSS Feed urls that are added to the Social Icons widget are correctly recognised and displayed using the RSS feed icon rather than the default link icon.

rickcurran committed:
Added more common feed url components to get_supported_icons()

As an improvement based on the issue #10662 "Social Icons Widget: allow for alternate feed URLs" this adds some more common feed url components to the get_supported_icons() function. The existing version only matches the existence of "/feed/" in the url being added, as such many other feed URL formats would end up showing with the default link icon instead. Whilst it may still not catch all possilbe feed url formats it will catch many commonly used formats as well as non-permalink versions of WordPress feeds.

jeherve commented on modules/widgets/social-icons.php:
```diff
@@ -512,6 +512,106 @@ public function get_supported_icons() {
+				'icon'  => 'feed',
+				'label' => __( 'RSS Feed', 'jetpack' ),
+			),
+            array(
```
You seem to have some spaces in there instead of tabs. I would recommend setting up your IDE as recommended here to get recommendations to solve all those errors at once:
https://github.com/Automattic/jetpack/blob/master/docs/development-environment.md#use-php-codesniffer-and-eslint-to-make-sure-your-code-respects-coding-standards

rickcurran replied:
Ack, sorry, I'll fix that!

rickcurran replied:
I've pushed a fix for those spaces instead of tabs now.

jeherve commented on modules/widgets/social-icons.php:
```diff
@@ -512,6 +512,106 @@ public function get_supported_icons() {
+				'label' => __( 'RSS Feed', 'jetpack' ),
+			),
+            array(
+				'url'   => '/.rss', // Yep, Reddit uses this
```
I think those types of comments are helpful, it helps us see at a glance what sites we support. Do you think you could add some more for the other formats in there?

rickcurran replied:
Sure, I'll see what other ones I can provide comments for. The ones I commented were very specific sources, the others aren't necessarily from one particular source of RSS but I'll see what I can add.

rickcurran replied:
I've added some comments for most of the RSS formats now, a few of them are unknown / generic ones I've come across so not specifically with any particular CMS / platform.

jeherve commented on modules/widgets/social-icons.php:
```diff
@@ -512,6 +512,106 @@ public function get_supported_icons() {
 				'icon'  => 'feed',
 				'label' => __( 'RSS Feed', 'jetpack' ),
 			),
+			array(
```
It's a bit annoying that we have so many entries for just one thing: RSS. I wonder if there would be a way to shorten this a bit by having one big array of potential RSS formats that we would then add all at once to this existing list?

rickcurran replied:
Yes, it is a bit unwieldy to have so many! Do you mean put all the RSS parameters into an array in the 'url', e.g:

`array(
    'url'   => array( '/feed/', '/feeds/', 'format=RSS', '/rss' ),
    'icon'  => 'feed',
    'label' => __( 'RSS Feed', 'jetpack' ),
)`

If so then it would require changing the way the urls are matched when the widget is rendered, currently it uses `if ( false !== stripos( $icon['url'], $social_icon['url'] ) ) {` within a foreach loop to see if any of the social_icon urls match the URL to be rendered, so I think we'd have to check to see if the `$icon['url']` was an array or not, then match inside that array if it is. Does that seem a viable way?

jeherve replied:
Yes, you could even pull that array of feed URL formats out of there, in a new `$feed_url_formats` array, and only add it later to the `$social_links_icons` array, right before it's returned, all inside
`get_supported_icons()`.

rickcurran replied:
Hi, I've pushed some changes along those lines now. I added a new `$feed_url_formats` array to hold the 'feed' url formats. I've then updated the code that renders the icons to check if the 'url' is an array and if so then to loop over the array and match inside that, if not an array then it matches as normal. 

This does mean there are two bits of identical code echoing the icon out which I'm not too happy about, but in the interest of not over-optimising the code I've gone with that for now. Happy for feedback / suggestions on better ways that could be handled if you think that needs more attention.

rickcurran committed:
Corrected spaces to tabs, clarified some of the RSS feed sources

Fixed incorrect use of spaces instead of tabs.
Added some additional comments regarding which sources RSS feeds may come from.

rickcurran committed:
Moved Feed formats into $feed_url_formats array

There are many possible RSS Feed url formats that can be used to match for the RSS Feed icon, so to reduce repetition of multiple 'feed' entries in the `$social_links_icons` array a separate `$feed_url_formats` now contains the common 'feed' icon and the various feed url parameters. As such it means the 'url' of an item in the `$social_links_icons` array can itself be an array so the code for rendering the widget has been updated to accommodate this.

jeherve commented on modules/widgets/social-icons.php:
```diff
@@ -137,11 +137,27 @@ public function widget( $args, $instance ) {
-										break;
+									
+									// Check if the $social_icon['url'] is an array ('Feed' url is provided as an array of url components)
+									if ( is_array( $social_icon['url'] ) ) {
```
That's a good approach. Would it be possible to simplify things even more by always having an array, but having an array with a single value in those cases? This way we would simplify things here.

rickcurran replied:
Ah yep, that would be definitely be a good way to simplify it and avoid the repetition!

rickcurran replied:
Ok, I've made that change now. I realised as I was committing that change that this means that the ten different 'amazon.xx' entries (amazon.cn, amazon.in, amazon.fr etc) in the `$social_links_icons` array could now be reduced to one with all the domain variations in that single entry.

jeherve replied:
> this means that the ten different 'amazon.xx' entries (amazon.cn, amazon.in, amazon.fr etc) in the $social_links_icons array could now be reduced to one with all the domain variations in that single entry.

Let's do it! :)

rickcurran replied:
Ok, done. I've committed a change that reduces Amazon, Google, Discord and WordPress domains into single entries instead.

rickcurran committed:
Simplified the code by setting all `$social_links_icons` urls as arrays

The previous commit added the various `$social_links_icons` 'feed' items as an array. This cleaned up the code but also added a bit of repetition due to having to check whether a `$social_links_icons` url element was an array and subsequently handle them in two different ways but duplicating some code in the process.

This update changes all url elements in the `$social_links_icons` array to be an array even if there is only one item (as most of them do). This means that the code used to check if a url matches only needs to presume the url element comes as an array. This also enables the possibilty that some of the existing items e.g. Amazon can be reduced to one entry in the `$social_links_icons` array.

rickcurran committed:
Reduced multiple entries for Amazon & others urls in `$social_links_icons` array

This commit reduces the separate entries that exist for Amazon, Google, Discord and WordPress domains into single items with multiple url array items instead.

jeherve committed:
Widgets: merge Feed array with rest of supported icons

Now that all supported services use the array format, we can move this whole array back in without creating any confusion.

jeherve committed:
Simplify the display of each icon

- Since we know for sure $social_icon['url'] is going to be an array, let's remove the extra check
- Use printf instead of 2 echo
- Add phpcs:ignore to avoid error on output

jeherve committed:
Social Icons: fix all PHPCS errors and warnings, & add to whitelist

jeherve reviewed (dismissed):
This tests well for me now! I've pushed a few commits to clear out a few things, but all in all I think this is good to go!

rickcurran commented:
> This tests well for me now! I've pushed a few commits to clear out a few things, but all in all I think this is good to go!

I had wondered about putting the 'feeds' array back into the main `$social_links_icons` array, thanks for doing that!

mdawaffe commented on modules/widgets/social-icons.php:
```diff
@@ -388,277 +416,256 @@ public function get_svg_icon( $args = array() ) {
+				'url'   => array(
+					'/feed/',        // WordPress default feed url.
+					'/feeds/',       // Blogspot and others.
+					'/blog/feed',    // No trailing space WordPress feed, could use /feed but may match unexpectedly.
```
Should this say "no trailing slash"?

Also, do we have a concrete example of where this entry would be useful?

rickcurran replied:
Yep, that should be "No trailing slash".

jeherve replied:
done in b82da9ec4ddb0ca3198fa52be6fbdae6ae207999

mdawaffe commented on modules/widgets/social-icons.php:
```diff
@@ -388,277 +416,256 @@ public function get_svg_icon( $args = array() ) {
+					'?feed=rss',     // WordPress non-permalink - Catches feed=rss / feed=rss2.
+					'?feed=rdf',     // WordPress non-permalink.
+					'?feed=atom',    // WordPress non-permalink.
+					'http://feeds',  // FeedBurner.
```
Would `http://feeds.` (note the dot) be better here?

rickcurran replied:
I think that would be a safe change to make, it is primarily intended to catch Feedburner feed urls so they would all be `http://feeds.`

jeherve replied:
done in b82da9ec4ddb0ca3198fa52be6fbdae6ae207999

mdawaffe commented on modules/widgets/social-icons.php:
```diff
@@ -388,277 +416,256 @@ public function get_svg_icon( $args = array() ) {
+					'/rss',          // Tumblr.
+					'/.rss',         // Reddit.
+					'/rss.xml',      // Moveable Type, Typepad.
+					'http://rss',
```
Would `http://rss.` (note the dot) be better here?

Also, do we have a concrete example of where this entry would be useful?

rickcurran replied:
I found it in use in a couple of feeds in amongst my RSS feed subscriptions, one url no longer works but `http://rss.macworld.com` still active – admittedly it does redirect to a different feed url when you access it.

jeherve replied:
done in b82da9ec4ddb0ca3198fa52be6fbdae6ae207999

mdawaffe reviewed (commented):
I like the comments explaining why each entry is useful. Do the rest of the entries have an explanation?

jeherve committed:
Clarify comments for RSS list.

See:
- https://github.com/Automattic/jetpack/pull/13183/files#r314844476
- https://github.com/Automattic/jetpack/pull/13183/files#r314844856
- https://github.com/Automattic/jetpack/pull/13183/files#r314844732

jeherve reviewed (approved):
LGTM.
